### PR TITLE
feature(CameraUtils): add Extent to transformCameraToLookAtTarget params

### PR DESCRIPTION
**Pull request by mgermerie**
## Description

- Add the possibility to pass an `Extent` as `params` in `CameraUtils.transformCameraToLookAtTarget(view, camera, params)` method. If so, the camera will be placed in a way that the entire given extent can be seen.
- Adapt `PlanarView` and `GlobeView` constructors so that they should be able to call `transformCameraToLookAtTarget` with an extent.
- The `pacement` parameter for `GlobeView` and the optional `options.placement` parameter for `PlanarView` can now be `Extents`.

### CODE REVIEW

**Comment by mgermerie** [comment]
This PR follows the work done in PR #1563 and shall be rebased after PR #1563 is merged.

**Comment by mbredif** [comment]
nice addition!
Beware of the typo though : it is FRUSTUM not FRUST*R*UM
https://threejs.org/docs/#api/en/math/Frustum

**Comment by mgermerie** [comment]
> nice addition!
> Beware of the typo though : it is FRUSTUM not FRUST_R_UM
> https://threejs.org/docs/#api/en/math/Frustum

Thanks ! I corrected the typo.

**Comment by mbredif** [comment]
I believe that your automatic axis selection is not what you want exactly (unless the view camera as a square viewport).
you should compare the aspect ratio of the extent and the one of the camera to decide which axis should be extended .

**Comment by mgermerie** [comment]
> I believe that your automatic axis selection is not what you want exactly (unless the view camera as a square viewport).
> you should compare the aspect ratio of the extent and the one of the camera to decide which axis should be extended .

Indeed, thanks for the advice ! I fixed the automatic axis selection.
Now, if extent's ratio is bigger than camera's, the camera frustum width will be equal to the extent `x` dimension. On another hand, if the extent's ratio is lower than camera's, the camera frustum height will be equal to the extent `y` dimension.
In each case, the other dimension of the camera frustum is computed from the first one and the camera ratio.

File: examples/misc_orthographic_camera.html [hunk 1]
```diff
@@ -43,14 +43,16 @@
...
             );
 
+
             // `viewerDiv` will contain iTowns' rendering area (`<canvas>`)
             const viewerDiv = document.getElementById('viewerDiv');
 
             // instantiate PlanarView with an orthographic camera
             const view = new itowns.PlanarView(viewerDiv, extent, {
-                camera: { type: itowns.CAMERA_TYPE.ORTHOGRAPHIC },
+                // camera: { type: itowns.CAMERA_TYPE.ORTHOGRAPHIC },
```
Comment (gchoqueux) [inline examples/misc_orthographic_camera.html]: typo?

File: examples/misc_orthographic_camera.html [hunk 2]
```diff
@@ -43,14 +43,16 @@
...
             });
             setupLoadingScreen(viewerDiv, view);
+            const subextent = new itowns.Extent('TMS:3946', 3, 1, 1).as('EPSG:3946');
```
Comment (gchoqueux) [inline examples/misc_orthographic_camera.html]: viewedExtent?

File: examples/view_3d_map.html [hunk 3]
```diff
@@ -110,6 +110,11 @@
             view.addEventListener(itowns.GLOBE_VIEW_EVENTS.GLOBE_INITIALIZED, function () {
                 // eslint-disable-next-line no-console
                 console.info('Globe initialized');
+                // const subextent = new itowns.Extent('TMS:4326', 19, 119808, 531060).as('EPSG:4326');
+
+                const subextent = new itowns.Extent('TMS:3857', 17, 45086, 66381).as('EPSG:4326');
```
Comment (gchoqueux) [inline examples/view_3d_map.html]: viewedExtent?

File: src/Controls/GlobeControls.js [hunk 4]
```diff
@@ -1164,7 +1166,7 @@ class GlobeControls extends THREE.EventDispatcher {
      * @return     {Promise}  A promise that resolves when transformation is complete
      */
     lookAtCoordinate(params = {}, isAnimated = this.isAnimationEnabled()) {
-        if (params.zoom) {
+        if (params.zoom && !params.isExtent) {
```
Comment (gchoqueux) [inline src/Controls/GlobeControls.js]: `params.tilt !== undefined` must not be within its `params.isExtent` scope

File: src/Utils/CameraUtils.js [hunk 5]
```diff
@@ -370,19 +372,81 @@ export default {
...
+        // setup camera near and far planes
+        camera.near = view.isGlobeView ? Math.max(15.0, 0.000002352 * ellipsoidSizes.x) : 0.1;
+        if (camera.isOrthographicCamera) {
+            camera.far = 2000;
+        } else if (view.isPlanarView) {
+            const dimensions = params.isExtent ? params.dimensions() : view.extent.dimensions();
+            camera.far = 200 * Math.max(dimensions.x, dimensions.y);
+        } else {
+            camera.far = ellipsoidSizes.x * 10;
+        }
```
Comment (gchoqueux) [inline src/Utils/CameraUtils.js]: could remove from here? and move initial Class?

File: src/Core/Prefab/PlanarView.js [hunk 6]
```diff
@@ -29,40 +28,31 @@ class PlanarView extends View {
...
         this.isPlanarView = true;
-
-        // Configure camera
-        const dim = extent.dimensions();
-        const max = Math.max(dim.x, dim.y);
-        const camera3D = this.camera.camera3D;
-        camera3D.near = 0.1;
-        camera3D.far = 2 * max;
-        this.camera.camera3D.updateProjectionMatrix();
+        this.extent = extent;
```
Comment (gchoqueux) [inline src/Core/Prefab/PlanarView.js]: could you remove?

File: src/Core/Prefab/PlanarView.js [hunk 7]
```diff
@@ -29,40 +28,31 @@ class PlanarView extends View {
...
-
-        CameraUtils.transformCameraToLookAtTarget(this, camera3D, placement);
+        // Configure camera
+        const placement = options.placement || extent;
+        if (!placement.isExtent) {
+            placement.coord = placement.coord || extent.center();
+            placement.tilt = placement.tilt || 90;
+            placement.heading = placement.heading || 0;
+            const dim = extent.dimensions();
+            placement.range = placement.range || Math.max(dim.x, dim.y);
```
Comment (gchoqueux) [inline src/Core/Prefab/PlanarView.js]: could keep  `placement.range = placement.range || max;`

File: src/Utils/CameraUtils.js [hunk 8]
```diff
@@ -370,19 +372,81 @@ export default {
...
+
+                // setup range for camera placement
+                const verticalFOV = THREE.Math.degToRad(camera.fov);
+                if (dimensions.x / dimensions.y > camera.aspect) {
+                    const focal = (view.domElement.clientHeight * 0.5) / Math.tan(verticalFOV * 0.5);
+                    const horizontalFOV = 2 * Math.atan(view.domElement.clientWidth * 0.5 / focal);
+                    cameraTransformOptions.range = dimensions.x / (2 * Math.tan(horizontalFOV * 0.5));
+                } else {
+                    cameraTransformOptions.range = dimensions.y / (2 * Math.tan(verticalFOV * 0.5));
+                }
```
Comment (gchoqueux) [inline src/Utils/CameraUtils.js]: could add method `getTransformFromExtent()`

File: src/Controls/GlobeControls.js [hunk 9]
```diff
@@ -273,10 +273,14 @@ class GlobeControls extends THREE.EventDispatcher {
...
+        if (placement.isExtent) {
+            positionObject(placement.center().as('EPSG:4978', xyz), cameraTarget);
+        } else {
+            positionObject(placement.coord.as('EPSG:4978', xyz), cameraTarget);
 
-        placement.tilt = placement.tilt || 89.5;
-        placement.heading = placement.heading || 0;
+            placement.tilt = placement.tilt || 89.5;
+            placement.heading = placement.heading || 0;
+        }
```
Comment (gchoqueux) [inline src/Controls/GlobeControls.js]:
```js
if (placement.isExtent) {
    placement.center().as('EPSG:4978', xyz);
} else {
    placement.coord.as('EPSG:4978', xyz);

	placement.tilt = placement.tilt || 89.5;
    placement.tilt = placement.tilt || 89.5;
	placement.heading = placement.heading || 0;
    placement.heading = placement.heading || 0;
}

positionObject(xyz);

```

File: src/Core/Prefab/PlanarView.js [hunk 10]
```diff
@@ -48,21 +44,23 @@ class PlanarView extends View {
...
-        CameraUtils.transformCameraToLookAtTarget(this, camera3D, placement);
+        // Configure camera
+        const placement = options.placement || extent;
+        if (!placement.isExtent) {
+            placement.coord = placement.coord || extent.center();
+            placement.tilt = placement.tilt || 90;
+            placement.heading = placement.heading || 0;
+            placement.range = placement.range || max;
+        }
+        CameraUtils.transformCameraToLookAtTarget(this, this.camera.camera3D, placement);
```
Comment (gchoqueux) [inline src/Core/Prefab/PlanarView.js]: just `camera3D`

File: src/Utils/CameraUtils.js [hunk 11]
```diff
@@ -370,23 +372,87 @@ export default {
...
-     * @param      {CameraUtils~CameraTransformOptions}  params  The parameters
+     * @param      {CameraUtils~CameraTransformOptions|Extent}  params  The parameters
      * @return     {Promise} promise with resolve final CameraUtils~CameraTransformOptions
      */
     transformCameraToLookAtTarget(view, camera, params = {}) {
-        params.proxy = params.proxy === undefined || params.proxy;
+        let cameraTransformOptions;
+
+        if (params.isExtent) {
+            cameraTransformOptions = this.getCameraTransformOptionsFromExtent(view, camera, params);
```
Comment (gchoqueux) [inline src/Utils/CameraUtils.js]: use params instead of cameraTransformOptions

**Review by gchoqueux — APPROVED** [review]

**Pull request merged**